scylla-cdc-replicator/src/replicator_consumer.rs

```diff
@@ -0,0 +1,244 @@
+    }
+
+    async fn update(&self, data: CDCRow<'_>) {
+        self.update_or_insert(data, false).await;
+    }
+
+    async fn insert(&self, data: CDCRow<'_>) {
+        self.update_or_insert(data, true).await;
+    }
+
+    async fn update_atomic_or_frozen<'a>(
+        &self,
+        column_name: &str,
+        data: &'a CDCRow<'_>,
+        mut values_for_update: Vec<&'a CqlValue>,
+        values_for_delete: &Vec<&CqlValue>,
+        timestamp: i64,
+    ) {
+        if let Some(value) = data.get_value(column_name) {
+            // Order of values: ttl, inserted value, pk condition values.
+            values_for_update[1] = value;
+            self.run_statement(
+                Query::new(format!(
+                    "UPDATE {}.{} USING TTL ? SET {} = ? WHERE {}",
+                    self.dest_keyspace_name, self.dest_table_name, column_name, self.keys_cond
+                )),
+                &values_for_update,
+                timestamp,
+            )
+            .await;
+        } else if data.is_value_deleted(column_name) {
+            self.run_statement(
+                Query::new(format!(
+                    "DELETE {} FROM {}.{} WHERE {}",
+                    column_name, self.dest_keyspace_name, self.dest_table_name, self.keys_cond
+                )),
+                &values_for_delete,
+                timestamp,
+            )
+            .await;
+        }
+    }
+
+    async fn update_or_insert(&self, data: CDCRow<'_>, is_insert: bool) {
+        let keys_iter = ReplicatorConsumer::get_keys_iter(&self.table_schema);
+        let ttl = CqlValue::BigInt(data.ttl.unwrap_or(0)); // If data is inserted without TTL, setting it to 0 deletes existing TTL.
+        let timestamp = ReplicatorConsumer::get_timestamp(&data);
+        let values = keys_iter
+            .clone()
+            .map(|col_name| data.get_value(col_name).as_ref().unwrap())
+            .collect::<Vec<&CqlValue>>();
+
+        if is_insert {
+            // Insert row with nulls, the rest will be done through an update.
+            self.run_prepared_statement(self.insert_query.clone(), &values, timestamp)
+                .await;
+        }
+
+        let mut values_for_update = Vec::with_capacity(2 + values.len());
+        values_for_update.extend([&ttl, &CqlValue::Int(0)]);
+        values_for_update.extend(values.iter());
+
+        for column_name in &self.non_key_columns {
+            match self.table_schema.columns.get(column_name).unwrap().type_ {
+                CqlType::Native(_)
+                | CqlType::Tuple(_)
+                | CqlType::Collection { frozen: true, .. }
+                | CqlType::UserDefinedType { frozen: true, .. } => {
+                    self.update_atomic_or_frozen(
+                        column_name,
+                        &data,
+                        values_for_update.clone(), // We can't pass a mutable ref here if we want to run async functions in a loop.
```

Comment:
This is weird. I swear that there was a compillation error saying that I can't use `&mut` because there can be only one mutable reference at a time, and one has been created in the previous iteration of the loop, but maybe it was related to some other bug that was removed in the meantime. Since it works, I'll just remove the comment and use `&mut`.